tests/integration/test_charm.py

```diff
@@ -17,33 +25,124 @@ async def juju_run(unit, cmd):
     return stdout
 
 
-async def test_build_and_deploy(ops_test: OpsTest):
+async def test_workload_online_default(ops_test: OpsTest, app: Application):
     assert ops_test.model
-    charm = await ops_test.build_charm(".")
-    await ops_test.model.deploy("postgresql-k8s")
-    await ops_test.model.deploy(charm)
-    await ops_test.model.wait_for_idle(raise_on_error=False)
-    await ops_test.model.add_relation(
-        "postgresql-k8s:db",
-        "mattermost-k8s",
+    mmost_unit = app.units[0]
+    action = await mmost_unit.run("unit-get private-address")
+    curl_output = await juju_run(
+        mmost_unit, "curl {}:8065".format(action.results["Stdout"].replace("\n", ""))
     )
-    # mypy doesn't see that ActiveStatus has a name
-    await ops_test.model.wait_for_idle(status=ActiveStatus.name, raise_on_error=False)  # type: ignore
+    assert "Mattermost" in curl_output
+
 
+async def test_s3_storage(
+    ops_test: OpsTest,
+    app: Application,
+    localstack_s3_config: dict,
+):
+    """
+    arrange: after charm deployed and openstack swift server ready.
+    act: update charm configuration for openstack object storage plugin.
+    assert: a file should be uploaded to the openstack server and be accessible through it.
+    """
 
-async def test_status(ops_test: OpsTest):
     assert ops_test.model
-    # mypy doesn't see that ActiveStatus has a name
-    assert ops_test.model.applications["mattermost-k8s"].status == ActiveStatus.name  # type: ignore
-    assert ops_test.model.applications["postgresql-k8s"].status == ActiveStatus.name  # type: ignore
 
+    await app.set_config(
+        {
+            "s3_enabled": "true",
+            "s3_endpoint": localstack_s3_config["url"],
+            "s3_bucket": localstack_s3_config["bucket"],
+            "s3_region": localstack_s3_config["region"],
+            "s3_access_key_id": localstack_s3_config["credentials"]["access-key"],
+            "s3_secret_access_key": localstack_s3_config["credentials"]["secret-key"],
+            "s3_server_side_encryption": "false",
+            "s3_tls": "false",
+        }
+    )
+    await ops_test.model.wait_for_idle(status="active")
 
-async def test_workload_online_default(ops_test: OpsTest):
-    assert ops_test.model
-    app = ops_test.model.applications["mattermost-k8s"]
-    mmost_unit = app.units[0]
-    action = await mmost_unit.run("unit-get private-address")
-    curl_output = await juju_run(
-        mmost_unit, "curl {}:8065".format(action.results["Stdout"].replace("\n", ""))
+    # create a user
+    cmd = "MMCTL_LOCAL_SOCKET_PATH=/mattermost/run/local.socket /mattermost/bin/mmctl --local user create --email test@test.test --username test --password thisisabadpassword"
+    output = await ops_test.juju("run", "--application", app.name, cmd)
+    print(output)
+
+    # login to the API
+    data = {"login_id": "test@test.test", "password": "thisisabadpassword"}
```

Comment:
Where did this password come from? Perhaps also better to use the secrets library and perhaps make the value a fixture so its shared?

Reply:
I created a fixture `test_user`